src/Main.elm

```diff
@@ -1,10 +1,829 @@
+                , E.width E.fill
+                ]
+                [ upperLeftBackButtonOrText
+                , E.el [ E.alignRight ] upperRightInfoElement
+                ]
+
+        browser : E.Element Msg
+        browser =
+            E.column
+                [ E.width E.fill
+                , E.height E.fill
+                , EBackground.color Palette.gray4
+                , E.spacing 1
+                , EBorder.rounded 7
+                , EBorder.width 1
+                , EBorder.color Palette.color5
+
+                -- we do this, otherwise the elements inside aren't going to respect
+                -- our rounded borders
+                , E.htmlAttribute <| Html.Attributes.style "overflow" "hidden"
+                ]
+                [ upperPanel
+                , E.column
+                    [ E.width E.fill
+                    , E.height E.fill
+                    , E.spacing 1
+                    ]
+                    -- folders first, just like "FIRST_STEP.md" says
+                    (filteredAudienceFolders ++ filteredAudiences)
+                ]
+    in
+    E.row
+        [ E.spacing 20
+        ]
+        [ E.el [ E.alignTop, E.width <| E.px 700 ] browser
+        , viewSortingPanel model.sortBy
+        ]
+
+
+viewFolder : Int -> FormattedAudienceFolder -> List (E.Element Msg) -> List (E.Element Msg)
+viewFolder id folder acc =
+    let
+        -- we view a "folder" as a row: the icon, and the name of the folder
+        item =
+            E.row
+                [ E.paddingXY 15 10
+                , EBackground.color Palette.gray5
+                , EEvents.onClick (FolderClicked id)
+                , EFont.color Palette.gray0
+                , E.width E.fill
+                , E.spacing 15
+                , E.htmlAttribute <| Html.Attributes.style "cursor" "pointer"
+                , E.mouseOver <| [ EBackground.color <| E.rgb255 255 255 255 ]
+                ]
+                [ E.el
+                    [ E.width <| E.px 18
+                    , E.height <| E.px 18
+                    ]
+                    <| E.html <| Icons.folder "#6ea4fc"
+                , E.text folder.name
+                ]
+    in
+    item :: acc
+
+
+viewAudience : Int -> FormattedAudience -> List (E.Element Msg) -> List (E.Element Msg)
+viewAudience _ audience acc =
+    let
+        -- we view an "audience" as a row: the icon, and the name of the audience
+        item =
+            E.row
+                [ E.paddingXY 15 10
+                , EBackground.color Palette.gray5
+                , E.width E.fill
+                , E.spacing 15
+                , EFont.color Palette.gray0
+                , E.mouseOver <| [ EBackground.color <| E.rgb255 255 255 255 ]
+                ]
+                [ E.el
+                    [ E.width <| E.px 17
+                    , E.height <| E.px 17
+                    ]
+                    <| E.html <| Data.Audience.audienceTypeToIcon audience.type_
```

Comment:
Just a curious question as you have more experience with elm-ui than me. How often do you find you need the E.html or E.htmlAttribute escape hatch? In my toy project I found I need it quite a lot for some layout tasks around eg. textareas etc... - elm-ui didn't allow me to flex-grow it as I wanted etc... Is this common?

Reply:
Yeah I almost never use `E.html`, except for APIs that return `Html.Html msg`.

## Cases where I've used `E.html`
- `Svg.svg`
- `WebGL.toHtml`

As for `E.htmlAttribute`, yeah I've actually used that a bit more.

## Notable cases where I've used `E.htmlAttribute`
- `E.htmlAttribute <| Html.Attributes.transform ...`
- `E.htmlAttribute <| Html.Attributes.style "z-index" ...`
- `E.htmlAttribute <| Html.Attributes.id ...`

Other than that I've actually done everything in elm-ui.

And especially when it comes to laying out text, there are certain "recipes" or guidelines that you use in elm-ui.
For example, to center text:
```elm

view model = E.layout [] <|
    E.el
        -- make this element take up the whole page width and height
        [ E.width E.fill
        , E.height E.fill
        ]
        <| E.el 
            -- since we don't specify here how much width or height this will take,
            -- it will take up only as much size as the text takes up, so it will
            -- be placed exactly in the middle.
            [ E.centerX
            , E.centerY
            ]
            <| E.text "This text will be centered."
```
In short, I've almost never really needed to use `E.html` or `E.htmlAttribute` to do layout, even when it comes to text.